### PR TITLE
ENH: add `cline` to `Styler.to_latex`

### LINKED ISSUE #21673: ENH: Styler.to_latex()
I have created a [branch](https://github.com/toobaz/pandas/tree/style_to_latex) adding a ``to_latex`` method to ``Styler``.

It is nowhere near readiness, and in particular:
- supports only cell background color as of now - adding some attributes should be trivial, adding others might be less so
- most importantly, it works on a copy of the dataframe casted ``.astype(str)``, which is plain wrong because it discards features such as float formatting. For a good result, a better integration is needed with the code that currently does cell formatting in ``DataFrame.to_latex``.

This said, if anyone feels like experimenting, it might be slightly better than starting from scratch.

One aspect which I think is useful, even aside from this branch, and might benefit some discussion, is the pair of methods ``_latex_preserve`` and ``_latex_restore``, which basically replace LaTeX commands so that they are not disturbed by escaping, and then restores them. There might be better way to code this, but I really think this is something we need to implement, and to offer to users who happen to nest LaTeX code in their cells.

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #21673

This allows `Styler.to_latex` to also render `\cline` to separate sparse MultiIndex rows according to user input. Tests are also included to demonstrate functionality with hidden rows and hidden index levels.

Note that DataFrame.to_latex also uses `\cline` but it is forced in certain cases without options.

```
midx = pd.MultiIndex.from_product([["A", "-", "B"], [0], ["X", "Y"]], names=["lvl 0", "lvl 1", "last lvl"])
df = pd.DataFrame([[1, "pig"], [2, "coin"], [99, ""], [99, ""], [3, "sky"], [4, "pie"]], 
                  index=midx, columns=["int", "str"])
styler = df.style
styler.hide([("-", 0, "X"), ("-", 0, "Y")])
styler.hide(level=[1,3])
print(styler.to_latex(clines="all;index", hrules=True, multirow_align="t"))
```

| | `all` | `skip-last` |
--- | --- | --- |
**`index`** | Screenshot 2021-12-31 at 14 49 53 | Screenshot 2021-12-31 at 14 55 22
**`data`** | Screenshot 2021-12-31 at 14 51 30 | Screenshot 2021-12-31 at 14 52 59 |

Default render without ``clines``:

Screenshot 2021-12-31 at 15 03 14

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -542,6 +543,22 @@ def to_latex(
             Defaults to ``pandas.options.styler.latex.hrules``, which is `False`.
 
             .. versionchanged:: 1.4.0
+        clines : str, optional
+            Use to control adding \\cline commands for the index labels separation.
+            Possible values are:
+
+              - `None`: no cline commands are added (default).
```
Comment (jreback) [inline pandas/io/formats/style.py:550]: can you add a validation test & error checking if the clines don't have the possible values

Comment (attack68) [reply pandas/io/formats/style.py:550]: done

File: pandas/io/formats/style_render.py [hunk 2]
```diff
@@ -789,6 +792,38 @@ def _translate_latex(self, d: dict) -> None:
             body.append(row_body_headers + row_body_cells)
         d["body"] = body
 
+        # clines are determined from info on index_lengths and hidden_rows and input
+        # to a dict defining which row clines should be added in the template.
+        if clines is not None and (
```
Comment (jreback) [inline pandas/io/formats/style_render.py:797]: can do
```clines not in [None, 'data', 'all', ....]```

also raise a ValueError

File: pandas/io/formats/style_render.py [hunk 3]
```diff
@@ -789,6 +792,38 @@ def _translate_latex(self, d: dict) -> None:
             body.append(row_body_headers + row_body_cells)
         d["body"] = body
 
+        # clines are determined from info on index_lengths and hidden_rows and input
```
Comment (jreback) [inline pandas/io/formats/style_render.py:795]: woudn't object to putting this into a separate function and calling here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style_render.py (hunk 3 above)
Comment (attack68) [reply pandas/io/formats/style_render.py:795]: took a look but the logic depends on a few vars created in the 10 lines above and a separate func would need to be passed a few vars, inc `self` which gets messier. think is better just to leave here as self contained block.

File: pandas/io/formats/style_render.py (hunk 2 above)
Comment (attack68) [reply pandas/io/formats/style_render.py]: done.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**